server/src/main/java/io/crate/metadata/cluster/AlterTableClusterStateExecutor.java

```diff
@@ -153,149 +150,45 @@ protected ClusterState execute(ClusterState currentState, AlterTableRequest requ
-        //
-        // _meta format:
-        //   {
-        //       primary_keys: [],
-        //       partitioned_by: []     -- items are a tuple of (column_name, type_name)
-        //       constraints: {
-        //           not_null : [],
-        //       },
-        //       indices: {
-        //          <index_name>: <options>
-        //      }
-        //       check_constraints: {
-        //           <constraint_name>: <expression>
-        //       }
-        //   }
-        //
-        //   - partitioned_by cannot be changed
-        //
-        //   - primary_keys are always additive
-        //      - If present in delta, ALL keys are present (We should change this to gain atomicity)
-        //
-        //   - not_null constraints are always additive
-        //      - If present in delta, NEW columns are present
-        //
-        //   - indices can only be added
-        //      - If present in delta, NEW indices are present
-        //
-        //   - check_constraints can be added OR removed
-        //      If removed, delta contains *ALL* but the one to remove
-        //      If added:
-        //          Empty if new column doesn't have a constraint
-        //          All constraints present if new column has a constraint
-        //          (We should change this to gain atomicity)
-        //
-        // Why is everything behaving differently? Because reasons 🤷
-        // This would be simpler if we distinguished between additions and deletions in the AlterTableRequest
-        Map<String, Object> mappingDeltaAsMap = request.mappingDeltaAsMap();
-        Map<String, Object> metaDelta = (Map<String, Object>) mappingDeltaAsMap.get("_meta");
-        if (metaDelta != null) {
-            var curPartitionedBy = currentMeta.get("partitioned_by");
-            if (curPartitionedBy != null) {
-                metaDelta.put("partitioned_by", curPartitionedBy);
-            }
-            var curPrimaryKeys = currentMeta.get("primary_keys");
-            if (curPrimaryKeys != null) {
-                metaDelta.putIfAbsent("primary_keys", curPrimaryKeys);
-            }
 
-            Map<String, Object> checkConstraints = (Map<String, Object>) metaDelta.get("check_constraints");
-            if (checkConstraints == null || checkConstraints.isEmpty()) {
-                var curCheckConstraints = currentMeta.get("check_constraints");
-                if (curCheckConstraints != null) {
-                    metaDelta.put("check_constraints", curCheckConstraints);
-                }
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+            if (indexMapping.get(ES_MAPPING_NAME).equals(mappingDelta.get(ES_MAPPING_NAME))) {
+                // This is important, if nothing has changed MapperService throws an error on the metadata update attempt
+                return currentState;
             }
+            indexMapping.put(ES_MAPPING_NAME, mappingDelta.get(ES_MAPPING_NAME));
```

Comment:
Isn't this always false as you compare a complete mapping vs. a delta only?
I could not find any code inside the MapperService.merge() path which would fail if nothing changed, could you please point me to this (if true, adding it to the comment would also be helpful) ?

Reply:
I'm referring to https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L236

>compare a complete mapping vs. a delta only?

I'm comparing concrete property which I'm going to change - as in this case we know it will be the only change in metadata. I just replace, say, strict to strict, MapperService will check prev version, note that nothing has changed and will throw an error 

jenkins [failure](https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/6365/testReport/junit/io.crate.integrationtests/ColumnPolicyIntegrationTest/testAlterDynamicTable/) before adding this extra check.

Pushed a [comment](https://github.com/crate/crate/pull/13708/commits/346cfbfc6f540967537e65627a35d15fde6323a7) to clarify the behavior.

Reply:
it's BTW somehow similar to https://github.com/crate/crate/pull/13734, skip to no-op and return existing cluster state if we know that change will not happen